maven-cli/src/main/java/org/apache/maven/cling/invoker/Utils.java

```diff
@@ -144,4 +150,37 @@ public static int toPlexusLoggingLevel(Slf4jConfiguration.Level level) {
             case ERROR -> Logger.LEVEL_ERROR;
         };
     }
+
+    @Nullable
+    public static Path findRoot(Path topDirectory) throws ParserException {
+        requireNonNull(topDirectory, "topDirectory");
+        Path rootDirectory = null;
+        for (RootLocator rootLocator : ServiceLoader.load(RootLocator.class).stream()
+                .map(ServiceLoader.Provider::get)
+                .toList()) {
+            rootDirectory = rootLocator.findRoot(topDirectory);
+            if (rootDirectory != null) {
+                break;
+            }
+        }
+        if (rootDirectory != null) {
+            return getCanonicalPath(rootDirectory);
+        }
+        return null;
+    }
+
+    @Nonnull
+    public static Path findMandatoryRoot(Path topDirectory) throws ParserException {
+        requireNonNull(topDirectory, "topDirectory");
+        Path rootDirectory;
+        for (RootLocator rootLocator : ServiceLoader.load(RootLocator.class).stream()
+                .map(ServiceLoader.Provider::get)
+                .toList()) {
+            rootDirectory = rootLocator.findMandatoryRoot(topDirectory);
```

Comment:
In this case, we'll take only the first `RootLocator`, because calling `findMandatoryRoot` should throw an exception if that `RootLocator` cannot find the root.  That's not really in line with the `findRoot` (not mandatory above).
Should that be:
```
return Optional.ofNullable(findMandatory(topDirectory)).orElseThrow();
```